### PR TITLE
Lock on creating redis connections

### PR DESCRIPTION
**Pull request by JunTaoLuo**
#253

### CODE REVIEW

**Comment by Tratcher** [comment]
@JonCole

File: src/Microsoft.Extensions.Caching.Redis/RedisCache.cs [hunk 1]
```diff
@@ -34,6 +35,8 @@ public class RedisCache : IDistributedCache, IDisposable
         private readonly RedisCacheOptions _options;
         private readonly string _instance;
 
+        private readonly SemaphoreSlim _connectionLock = new SemaphoreSlim(1, 1);
```
Comment (Tratcher) [inline src/Microsoft.Extensions.Caching.Redis/RedisCache.cs]: named parameters please

**Comment by JonCole** [comment]
@JunTaoLuo,  

I would recommend that you structure the fix something like the following psuedo-code to avoid lock contention when the connection has already been created.

```
if (connection is null)
{
    //avoid taking the lock in the cases where the connection 
   //has already been created.
    TakeLock();  
    {
        if(connection is null)
        {
              CreateConnection()
        }
    }     
}
```

**Comment by JunTaoLuo** [comment]
@JonCole Yup that works, I was just talking to @sebastienros about the exact same thing.

**Comment by JunTaoLuo** [comment]
@muratg we need to decide whether this goes into 1.0.4 as well.

cc @mikeharder

**Comment by muratg** [comment]
@JunTaoLuo Please file a separate bug for 1.0.4. (and link it from #253).

Also, we should keep #253 open after this is checked in to dev branch.  For 1.1.1, we'll need to port it to some rel/* branch as 1.1.1 approaches.

**Comment by mikeharder** [comment]
The code looks correct.  Personally, I would extract the initialization logic into a helper class, named something like `AsyncLazy<T>`.  Similar to `Lazy<T>` in the framework , but supporting both sync and async initialization.  Then the code in this class would just be the following:

```
private AsyncLazy<IDatabase> _cache = new AsyncLazy<IDatabase>(() => /* sync init code */, () => /* async init code */);

var cache = _cache.Value();
var cache2 = await _cache.AsyncValue();
```

You shouldn't even need to keep an explicit reference to the ConnectionMultiplexer, since the IDatabase reference will prevent it from being GC.  This change would also eliminate the need to call `Connect()` before other operations.

File: src/Microsoft.Extensions.Caching.Redis/RedisCache.cs [hunk 2]
```diff
@@ -28,12 +29,14 @@ public class RedisCache : IDistributedCache, IDisposable
         private const string DataKey = "data";
         private const long NotPresent = -1;
 
-        private ConnectionMultiplexer _connection;
+        private volatile ConnectionMultiplexer _connection;
```
Comment (davidfowl) [inline src/Microsoft.Extensions.Caching.Redis/RedisCache.cs]: Why is this volatile?

**Pull request merged**